pandas/tests/indexes/datetimes/test_tools.py

```diff
@@ -469,6 +472,40 @@ def test_to_datetime_tz(self, cache):
         with pytest.raises(ValueError, match=msg):
             pd.to_datetime(arr, cache=cache)
 
+    @pytest.mark.parametrize('cache', [True, False])
+    def test_to_datetime_offset(self, cache):
+        # inspired by asv timeseries.ToDatetimeNONISO8601 benchmark
+        # see GH-26097 for more
+        ts_string_1 = 'March 1, 2018 12:00:00+0400'
+        ts_string_2 = 'March 1, 2018 12:00:00+0500'
+        arr = [ts_string_1] * 5 + [ts_string_2] * 5
+        expected = pd.Index([parse(x) for x in arr])
+        result = pd.to_datetime(arr, cache=cache)
+        tm.assert_index_equal(result, expected)
+
+    @pytest.mark.parametrize('errors', ('ignore', 'coerce', 'raise'))
+    @pytest.mark.parametrize('suffix', ([], ['foo']))
+    @pytest.mark.parametrize('convertor', (lambda x: x, str))
+    @given(date1=datetimes(timezones=timezones()),
+           date2=datetimes(timezones=timezones()))
+    def test_to_datetime_cache_errors(self, date1, date2, suffix,
+                                      errors, convertor):
+        arg = [convertor(date1), convertor(date2)] * 5 + suffix
+
+        def _get_answer(cache):
```

Comment:
you should not catch this error, rather for errors='raise' you use an assert_raises

Reply:
Thing is, `errors='raise'` does not guarantee it raises - that is, if `to_datetime` input is valid it won't raise.
What I wanted to verify with this test is that `to_datetime()` behaviour is identical regardless of whether it had `cache=True` or `cache=False`.

For this particular test scenario I don't even care if, say, `errors='ignore'` raises a `ValueError` as long as it is raised consistently with and without cache. It is up to other tests to verify that `errors` parameter is doing what it should, this test only checks for integrity over `cache`.

Reply:
I don't want a try/except in a test; that just obscures what the test is checking; as I said if you expect it to raise then make it explici.

Reply:
I tried re-working `try..except` away, but I feel that it's somewhat impossible unless I hardcode some of current Pandas behaviour in the test itself.

My current intentions are like this - `pd.to_datetime()` must behave the same when `cache=True` and `cache=False`, where "same" means "both return `pd.Index`" or "both raise _same_ exception".

So far I'm struggling with removal of `try..except` to cover this intention due to the fact that `pd.to_datetime` can raise an exception not only when `errors="raise"` (or it can _not_ raise an exception if hypothesis randomly generates equal timezones).

Example:

```python
import pandas as pd
from datetime import datetime as dt
import pytz
pd.to_datetime([dt(2000,1,1),dt(2000,1,1,tzinfo=pytz.utc)], errors='coerce')
```
results in:
```
Traceback (most recent call last):
  File "somewhere\pandas\core\arrays\datetimes.py", line 1850, in objects_to_datetime64ns
    values, tz_parsed = conversion.datetime_to_datetime64(data)
  File "pandas\_libs\tslibs\conversion.pyx", line 185, in pandas._libs.tslibs.conversion.datetime_to_datetime64
    raise ValueError('Cannot mix tz-aware with '
ValueError: Cannot mix tz-aware with tz-naive values

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "somewhere\pandas\util\_decorators.py", line 188, in wrapper
    return func(*args, **kwargs)
  File "somewhere\pandas\core\tools\datetimes.py", line 642, in to_datetime
    result = convert_listlike(arg, box, format)
  File "somewhere\pandas\core\tools\datetimes.py", line 333, in _convert_listlike_datetimes
    allow_object=True)
  File "somewhere\pandas\core\arrays\datetimes.py", line 1855, in objects_to_datetime64ns
    raise e
  File "somewhere\pandas\core\arrays\datetimes.py", line 1846, in objects_to_datetime64ns
    require_iso8601=require_iso8601
  File "pandas\_libs\tslib.pyx", line 460, in pandas._libs.tslib.array_to_datetime
    cpdef array_to_datetime(ndarray[object] values, str errors='raise',
  File "pandas\_libs\tslib.pyx", line 537, in pandas._libs.tslib.array_to_datetime
    raise ValueError('Tz-aware datetime.datetime '
ValueError: Tz-aware datetime.datetime cannot be converted to datetime64 unless utc=True
```

Reply:
> as I said if you expect it to raise then make it explici.

All that I said above is to illustrate that I actually _cannot_ expect it to raise or not to raise, as right now the behaviour of `pd.to_datetime()` regarding of exceptions can be described as "chaotic good" :)

Reply:
move the error='raise' case to another test then. we *never* do this in a test

Reply:
But what about exceptions which _might_ be raised when `errors != "raise"`? I feel this is overextending what the test is checking - when I describe what it must raise or that it musn't raise I start checking behaviour of `pd.to_datetime()` for certain parameters, while the only check I wanted is that it behaves _the same way_ with cache on/off.

Reply:
my point is that this code is very hard to grok when errors=‘raise’

I agree that both cache=True or False must raise in this case, but I want these in a separate test where you assert that they raise

the issue is that catching a ValueError in a test can accidentally catch other things and has multiple times in the past in some test cases which turned out badly

hence I don’t allow this ever

Reply:
Again, I get what you want, but I seem to fail to convey my own point.

What you describe is testing that `to_datetime()` behaves _correctly_ when given these or that parameter compositions. Those tests are needed, they're worthwhile, etc. But they're _not_ what I am adding here. What I want to test is something like "when given random parameters, `to_datetime()` must show identical results for `cache=True` and `cache=False`", so that turning caching on or off is merely a question of whether we have enough memory to speed up stuff (not a question of having a behaviour change)!

This check also verifies that potential bugs in the logic when it raises even if it shouldn't, or if it doesn't raise when it should, are treated equally by `cache`. _Finding_ those bugs is **not the purpose** of this particular test!
This includes e.g. cases when `to_datetime` raises an exception when `errors="coerce"`, and I don't think I should be trying to fix _all_ the errors in this function within this PR.

All in all, in this test I'm not interested ever if `to_datetime` does what it should. It might be shooting a rocket to the Moon for all I care, but it must be shooting same rocket to the same Moon both with cache on and off!
Or, more precisely here, it might be raising some random `ValueError`, but, unless it raises _different_ `ValueError`s for on/off cache, I consider it _fine_. Again, checking `to_datetime` correctness is _not the point_ of this particular test.

Reply:
All that said, would it be easier to understand if I extract this `_get_answer` function to be a module-level helper instead of a nested function (with a proper docstring to boot)?